Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -104,18 +108,48 @@ void WebPermissionController::tryProcessingRequests()
         }
 
         currentRequest.isWaitingForReply = true;
-        m_page->sendWithAsyncReply(Messages::WebPageProxy::QueryPermission(currentRequest.origin, currentRequest.descriptor), [this, weakThis = WeakPtr { *this }](auto state, bool shouldCache) {
-            RefPtr protectedThis { weakThis.get() };
-            if (!protectedThis)
-                return;
 
+        // If the context is provided in the request, the API being queried is Notifications and so the permission
+        // will be obtained by using the context rather then sending a message to the UIProcess
```

Comment:
This cannot be correct, I don't think any single WebContent process has a global view of notification permissions, only the UIProcess does. You need to IPC the UIProcess and get the permission from there.

Reply:
There is a Notification.permission, and it gets the state in web process

Reply:
I just looked at the code and it is true that we currently sync the notification permissions from the UIProcess to the WebProcess and that Notification.permission() doesn't currently query the UIProcess. I think this was needed because the Web API is synchronous.

However, the new Web API for permissions is async. I'll defer to @youennf who worked on this more but I still feel that the permission API should get all its permissions from the UIProcess instead of having bypasses in the WebProcess for specification permission types.

Note that even if @youennf does decide that getting the permission from the WebProcess is OK, the current behavior of relying on the presence of a context or not to determine if this is Notification permission is ugly and definitely not the right design.

Reply:
The goal of this patch is to add support for querying notification's permission. According to spec, the query() result is the same as Notification.permission, so I think the idea to invoke existing functions in web process is not wrong.

And we don't cache permission result in UI process now. I think we may want to move notification permission state to UI process to be align with other features, but that's another task. 

I agree with you that the implementation is wrong to make the decision based on whether context is nullptr.

Reply:
The UIProcess has the information via WebNotificationProvider::notificationPermissions().

Reply:
Also, if we really want to do this in the WebProcess, we don't need the ScriptExecutionContext at all AFAICT, you can just get the permission from `WebProcess::singleton().supplement<WebNotificationManager>()`.

Reply:
Where is it in UIProcess? Do you mean in WebKit's code? For permission like geolocation, it's stored in WebKit's clients. 

Yes I think we can use WebNotificationManager.

Reply:
I already said earlier but the UIProcess can get the information from `WebNotificationProvider::notificationPermissions()`. It will ask the client for the permissions.

Reply:
ohoh I thought by "UI process has the information" you mean WebKit in UI process caches the result. Yeah we may want to let web process pull the permission state from UI process instead of push the state to web process later.

Reply:
I've changed the implementation here so that notifications is not treated as an edge case.